File: binary_search/binary_search.go
package main

import "fmt"

// BinarySearch perform Binary Search by Iterative Method.
// Time Complexity : O(log(len(array)))
func BinarySearch(array []int, target int) int {
	left, right := 0, len(array)-1
	for left <= right {
		mid := (left + right) / 2
		if array[mid] == target {
			return mid
		}
		if array[mid] < target {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	return -1
}

func main() {
	array := []int{1, 5, 35, 112, 258, 324, 456, 512}
	index := BinarySearch(array, 112)
	if index == -1 {
		fmt.Println("Element is not present in array")
	} else {
		fmt.Println("Element is present at index :", index)
	}
}

File: binary_search/binary_search.cpp
// C++ program to implement recursive Binary Search
#include <bits/stdc++.h>
using namespace std;
//header file

int binarySearch(int arr[], int low, int high, int target)
{
    if (high >= low){
        int mid = low + (high-low)/2;//middle of the array
        if (arr[mid] == target)
            return mid;//If we find the tarrget element in the middle of the array
        else if(arr[mid] > target)//target element to left of the mid
            return binarySearch(arr, low, mid - 1, target);
        else if(arr[mid] < target)//target element to the right of the mid
            return binarySearch(arr, mid + 1, high, target);
    }
    return -1;//target eleement is not present in the array
}

int main()
{   int n;
    cout<<"Enter the size of array: ";
    cin>>n;
    int *arr= new int[n];
    cout<<"Enter the array(it should be in ascending order)"<<endl;
    for(int i=0;i<n;i++)
        cin>>arr[i];
    int x;
    cout<<"Enter the target element: ";
    cin>>x;
    int result = binarySearch(arr, 0, n - 1, x);
    (result == -1)
        ? cout << "Element is not present in array"
        : cout << "Element is present at index " << result;
    return 0;
}

//Time Complexity: O(log n)
//Space Complexity:O(log n)
